### ISSUE TITLE
BadageDrawable attached to the wrong menu item after making the item visible

### ISSUE THREAD

**Issue by Amrjyniat** [material-components/material-components-android#3569]
I have two menu items on the toolbar, the first item in the menu layout is hidden by default and becomes visible under some conditions, now I attached the BadageDrawable with the second menu item and it works fine, no problem.
The problem is the BadageDrawable is attached to the first menu item when becomes visible which isn't the intended item. then return attached the BadageDrawable with the second item after hiding the first item.

https://user-images.githubusercontent.com/61949013/191682188-9fa29ebd-1771-4dd1-a958-c5de18d0009c.mp4

### Reproduction steps
 - Create a menu layout with two items and hide the first item.
 - Attach the BD to the second item.
 - Show the first item on some conditions.

### Version number
 - 1.7.0-beta01

### Operating system and device
Windows 10, tested on Pixel API 27

**Comment by imhappi** [comment]
Unfortunately this looks like it is due to internal menu code: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/com/android/internal/view/menu/BaseMenuPresenter.java;drc=ba2e855cf782e13d751980a0ccc92b3d6ef2259d;l=83

We attach the badge to the menu item view based on the menu item id, but this view actually gets re-used

Not sure if there's anything we can do on our side, it may just need to be worked around by detaching and re-attaching badges when there are structural changes to the menu

**Comment by mobilekosmos** [comment]
Closing this as is, really? This issue makes the API actually really bad, I spent days to make it work just because of the unexpected behavior, if you for example have an 1 activity app where all fragment share the appbar, each time you navigate to another fragment that changes the appbar you will encounter this issue. At least talk to the guys of the internal menu code, or add this to the API documentation.
